### PR TITLE
Allow non-type converter instances in `ArrayConverter` and `HashValueConverter`

### PR DESCRIPTION
**Pull request by HertzDevil**
Resolves part of #10637. Also adds specs for these converters since for some reason there weren't any.

The old class methods simply delegate to a new converter instance (e.g. `ArrayConverter(T).from_json` -> `ArrayConverter::WithInstance(T.class).new(T).from_json`); this has negligible impact on execution time, and LLVM can probably optimize out the temporary struct. We could deprecate the class methods later if that makes the interface cleaner.

### CODE REVIEW

**Comment by straight-shoota** [comment]
Couldn't we just keep the class methods as is?

**Comment by asterite** [comment]
> Couldn't we just keep the class methods as is?

Aren't they the same? 🤔 

Though technically this is a breaking change because of going from `module` to `struct`, but I doubt someone is reopening those types.

**Comment by HertzDevil** [comment]
There is at least [one piece of code on GitHub](https://github.com/HCLarsen/ical_parser.cr/blob/2df2b1b3c67cbb0d7980259933e6d4a5b2d5c510/src/ical_parser/converters.cr) that repeats the standard library's `JSON::ArrayConverter` verbatim. I can't find any other examples of reopening that converter. In fact I can't find any *uses* of `JSON::HashValueConverter` on GitHub outside Crystal itself and the `json_mapping` shard.

Would absolutely no breaking changes like this PR and the `Nil` return type ones be allowed for the rest of 1.x? What should be the alternative solution if we want this PR not to be one of them?

**Comment by asterite** [comment]
> Would absolutely no breaking changes like this PR and the Nil return type ones be allowed for the rest of 1.x?

I think so

> What should be the alternative solution if we want this PR not to be one of them?

Introduce another type that provides the new features.

I personally don't like it, but I think that's the way forward if we want to avoid breaking changes.

**Comment by HertzDevil** [comment]
Any suggestions for the new type's name?

**Comment by asterite** [comment]
`ArrayConverterWithInstance` or something like that.

**Comment by straight-shoota** [comment]
I would see a possibility to merge such a change. If it changes someones code, that's because they're using the API not in the way it is intended. We probably won't be able to stay 100% clear of breaking changes anyways, because fixing bugs can always mean breaking things.

**Comment by asterite** [comment]
I think in this case it's probably fine to break things. I don't think these types are used a lot, as shown by that GitHub search. And we can always provide an upgrade path if needed, or actually help those shards in need. I have no idea why another shard would define those same types, maybe that's where the idea came from, or they couldn't wait for a release to have those types.

**Comment by HertzDevil** [comment]
What do you think about having a new class method `ArrayConverter.with_instance` that returns a converter of a private type, similar to what we do with most methods in `Iterator`?

> fixing bugs can always mean breaking things.

This is what I fear about; any bug fix can be considered a breaking change if any old code relies on buggy behaviour.

**Comment by straight-shoota** [comment]
`ArrayConverter.with_instance` could just be `ArrayConverter.new` and still return an instance of a private type.

But, I'd prefer merging the module to struct conversion.

**Comment by asterite** [comment]
Oh, adding `new` to the module is a great idea! I think it has the same effect, and it's not a breaking change.

We could change it to a `struct`, if really needed, for 2.0

File: src/json/from_json.cr [hunk 1]
```diff
@@ -391,26 +391,38 @@ struct Time::Format
...
 end
 
 module JSON::HashValueConverter(Converter)
-  def self.from_json(pull : JSON::PullParser)
-    hash = Hash(String, typeof(Converter.from_json(pull))).new
-    pull.read_object do |key, key_location|
-      parsed_key = String.from_json_object_key?(key)
-      unless parsed_key
-        raise JSON::ParseException.new("Can't convert #{key.inspect} into String", *key_location)
+  private struct WithInstance(T)
```
Comment (straight-shoota) [inline src/json/from_json.cr:410]: Minor, but I don't see a reason for the `With` prefix.

```suggestion
  private struct Instance(T)
```

**Review by straight-shoota — APPROVED** [review]

File: src/json/from_json.cr (hunk 1 above)
Comment (HertzDevil) [reply src/json/from_json.cr:410]: IMO `Instance` is confusing even if only used internally, because that has connotations of the singleton pattern.

**Pull request merged**